### PR TITLE
Add `LightningDataModule.load_from_checkpoint` to load datamodules directly from checkpoint

### LINKED ISSUE #12062: Datamodule's hparams should be saved inside checkpoint
## 🚀 Feature

LDM supports `save_hyperparameters` method just like LM.
https://github.com/PyTorchLightning/pytorch-lightning/blob/1ef34c5ffd20a42c687682798eb7db9b1dbcdb05/pytorch_lightning/core/datamodule.py#L26

but during checkpointing, only `model.hparams` are saved.
https://github.com/PyTorchLightning/pytorch-lightning/blob/1ef34c5ffd20a42c687682798eb7db9b1dbcdb05/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L365-L374

### Motivation

A user can have data specific hparams, like batch_size, image_size for transforms etc...
There can be cases where these hparams are required again, for eg in case of transfer learning/fine-tuning and user might be able to reload the hparams used while pre-training, without remembering what hparams were used during init, if they are saved inside the checkpoint.

Moreover, if we support this, we can possibly introduce `LDM.load_from_checkpoint` method just like we have in LM.

### Pitch

2 possible ways we can save LDM hparams:
1. Inject the LDM hparams inside model hparams, just like we do with `logger.log_hyperparams`
https://github.com/PyTorchLightning/pytorch-lightning/blob/1ef34c5ffd20a42c687682798eb7db9b1dbcdb05/pytorch_lightning/trainer/trainer.py#L1273
2. Create separate keys for each i.e. change `hyper_parameters` to `model_hyper_parameters` and add another `datamodule_hyper_parameters`.

### Alternatives

One can use `on_save_checkpoint/on_load_checkpoint` to reload the hparams, but this is slightly inconsistent with what we have within LM.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @awaelchli @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12062

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @awaelchli @borda @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
--
+- Add `LightningDataModule.load_from_checkpoint` to support loading datamodules directly from checkpoint ([#12550](https://github.com/PyTorchLightning/pytorch-lightning/pull/12550))
```
Comment (carmocca) [inline CHANGELOG.md:12]:
```suggestion
- Added `LightningDataModule.load_from_checkpoint` to support loading datamodules directly from checkpoint ([#12550](https://github.com/PyTorchLightning/pytorch-lightning/pull/12550))
```

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -262,3 +267,72 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
+                to the checkpoint.
+                However, if your checkpoint weights don't have the hyperparameters saved,
+                use this method to pass in a ``.yaml`` file with the hparams you'd like to use.
+                These will be converted into a :class:`~dict` and passed into your
+                :class:`LightningDataModule` for use.
+
+                If your datamodule's ``hparams`` argument is :class:`~argparse.Namespace`
+                and ``.yaml`` file has hierarchical structure, you need to refactor your datamodule to treat
+                ``hparams`` as :class:`~dict`.
+            \**kwargs: Any extra keyword args needed to init the datamodule. Can also be used to override saved
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:302]: I don't think this is required? At least we don't use it everywhere else

```suggestion
            **kwargs: Any extra keyword args needed to init the datamodule. Can also be used to override saved
```

File: pytorch_lightning/core/saving.py [hunk 3]
```diff
@@ -247,6 +170,111 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
 
 
+def _load_from_checkpoint(
+    cls: Union["pl.LightningModule", "pl.LightningDataModule"],
+    checkpoint_path: Union[str, IO],
+    map_location: Optional[Union[Dict[str, str], str, torch.device, int, Callable]] = None,
+    hparams_file: Optional[str] = None,
+    strict: Optional[bool] = None,
+    **kwargs,
+):
```
Comment (carmocca) [inline pytorch_lightning/core/saving.py:180]:
```suggestion
    **kwargs: Any,
) -> Any:
```

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -406,6 +407,16 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
             else:
                 checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_KEY] = dict(model.hparams)
 
+        if datamodule and datamodule.hparams:
+            if hasattr(datamodule, "_hparams_name"):
+                checkpoint[pl.LightningDataModule.CHECKPOINT_HYPER_PARAMS_NAME] = datamodule._hparams_name
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:412]: Shouldn't this be?

```suggestion
                checkpoint[datamodule.CHECKPOINT_HYPER_PARAMS_NAME] = datamodule._hparams_name
```

Also for `LightningModule`, in case it was modified

File: pytorch_lightning/core/saving.py [hunk 5]
```diff
@@ -247,6 +170,111 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
+    with pl_legacy_patch():
+        if map_location is not None:
+            checkpoint = pl_load(checkpoint_path, map_location=map_location)
+        else:
+            checkpoint = pl_load(checkpoint_path, map_location=lambda storage, loc: storage)
```
Comment (carmocca) [inline pytorch_lightning/core/saving.py:185]:
```suggestion
    if map_location is None:
        map_location = lambda storage, loc: storage
    with pl_legacy_patch():
        checkpoint = pl_load(checkpoint_path, map_location=map_location)
```

File: pytorch_lightning/core/saving.py [hunk 6]
```diff
@@ -247,6 +170,111 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
+    if hparams_file is not None:
+        extension = hparams_file.split(".")[-1]
+        if extension.lower() == "csv":
+            hparams = load_hparams_from_tags_csv(hparams_file)
+        elif extension.lower() in ("yml", "yaml"):
+            hparams = load_hparams_from_yaml(hparams_file)
+        else:
+            raise ValueError(".csv, .yml or .yaml is required for `hparams_file`")
+
+        hparams["on_gpu"] = False
```
Comment (carmocca) [inline pytorch_lightning/core/saving.py:196]: I know this is copy-pasted, but why do we set this key in particular to False? Where does this come from?

File: pytorch_lightning/core/saving.py [hunk 7]
```diff
@@ -247,6 +170,111 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
+
+        # overwrite hparams by the given file
+        checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = hparams
+
+    # for past checkpoint need to add the new key
+    if cls.CHECKPOINT_HYPER_PARAMS_KEY not in checkpoint:
+        checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = {}
```
Comment (carmocca) [inline pytorch_lightning/core/saving.py:203]:
```suggestion
    checkpoint.setdefault(cls.CHECKPOINT_HYPER_PARAMS_KEY, {})
```

File: pytorch_lightning/core/saving.py [hunk 8]
```diff
@@ -247,6 +170,111 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
+        return _load_state(cls, checkpoint, **kwargs)
+    return _load_state(cls, checkpoint, strict=strict, **kwargs)
+
+
+def _load_state(
+    cls: Union["pl.LightningModule", "pl.LightningDataModule"],
+    checkpoint: Dict[str, Any],
+    strict: Optional[bool] = None,
+    **cls_kwargs_new,
+):
```
Comment (carmocca) [inline pytorch_lightning/core/saving.py:217]:
```suggestion
    **cls_kwargs_new: Any,
) -> Any:
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/core/datamodule.py:302]: in wandb we do: https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.loggers.WandbLogger.html#wandblogger and it correctly reflects what `kwargs` are here for.

else it displays it like a normal argument.
https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#pytorch_lightning.core.lightning.LightningModule.load_from_checkpoint.params.kwargs

File: pytorch_lightning/core/saving.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/core/saving.py:196]: checked the git blame. Seems to be there since its inception. Should be removed since it's redundant now.

Comment (rohitgr7) [inline pytorch_lightning/core/saving.py:196]:
```suggestion
```

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 9]
```diff
@@ -396,15 +397,17 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
-                checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_NAME] = model._hparams_name
-            # dump arguments
-            if _OMEGACONF_AVAILABLE and isinstance(model.hparams, Container):
-                checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_KEY] = model.hparams
-                checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_TYPE] = type(model.hparams)
-            else:
-                checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_KEY] = dict(model.hparams)
+        for obj in (model, datamodule):
+            if obj and obj.hparams:
+                if obj.hparams:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:402]: This is already in the `if` above

**Review by kaushikb11 — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Review by awaelchli — COMMENTED** [review]
This is a nice feature addition!

I have one concern question where I believe we should look for a decision before merging the PR. 
In https://github.com/PyTorchLightning/pytorch-lightning/issues/10375 I propose to deprecate support for saving and restoring from a singular hparams container passed as input argument. If we decide to do this deprecation in this release, it would be better if the implementation here would already adopt the new restricted (simpler) way of parsing the args. If you think this is less convenient, then I'm also good with this. I just want to avoid us introducing a new feature with a broader scope and immediately deprecate partial functionality. 


cc @carmocca

**Comment by rohitgr7** [comment]
> This is a nice feature addition!
> 
> I have one concern question where I believe we should look for a decision before merging the PR. In #10375 I propose to deprecate support for saving and restoring from a singular hparams container passed as input argument. If we decide to do this deprecation in this release, it would be better if the implementation here would already adopt the new restricted (simpler) way of parsing the args. If you think this is less convenient, then I'm also good with this. I just want to avoid us introducing a new feature with a broader scope and immediately deprecate partial functionality.
> 
> cc @carmocca

let's do it in this release in a follow-up. I have seen similar issues related to it on slack so would be worth restricting the containers here.

**Review by Borda — APPROVED** [review]
lgtm

**Review by Borda — COMMENTED** [review]
.
